runtime/Cpp/runtime/src/atn/ArrayPredictionContext.cpp

```diff
@@ -22,6 +22,10 @@ namespace {
   }
 
   bool predictionContextEqual(const Ref<const PredictionContext> &lhs, const Ref<const PredictionContext> &rhs) {
+    if (lhs == nullptr)
```

Comment:
I wonder how badly this will hit performance

Reply:
> I wonder how badly this will hit performance

I think this question is difficult to answer without changing the Cpp runtime to not create "parent" arrays with null pointers. For tests that call the comparison function, the parse crashes; for tests that don't call the function, the parse works. Presumably, in Java when it calls [equals()](https://github.com/antlr/antlr4/blob/aa1f1f12a846846fcb78544bdb2ad135471376db/runtime/Java/src/org/antlr/v4/runtime/atn/ArrayPredictionContext.java#L78)] on comparing two "parents" arrays, it will test whether the element in the array is null (see [Arrays.equals()](https://github.com/openjdk/jdk/blob/0ee25de704eb8f179888812bd9b1e7bebc384fd6/src/java.base/share/classes/java/util/Arrays.java#L2979) => [Object.equals()](https://github.com/openjdk/jdk/blob/0ee25de704eb8f179888812bd9b1e7bebc384fd6/src/java.base/share/classes/java/util/Objects.java#L64)). This change makes Cpp operate the same way as the Java code. The choice is either to get the Cpp working, or rewrite all the runtimes to not use null pointers in "parents".

Reply:
I suspect it doesn't make any sense to have nulls in parents arrays, so I would opt for x-target cleanup, which as a side effect is likely to slightly improve performance

Reply:
I could swear that we had a sentinel node that represented top of the stack rather than null but I have to look at the code more carefully because @kaby76 found something that makes it look like it really is using null.  I am terrified of changing anything in the ATN simulation given just how much code in the world relies on it.

Reply:
The old code compare called [Arrays::equals()](https://github.com/antlr/antlr4/blob/53b83e329b67e7095a37bc155e172aea79f41558/runtime/Cpp/runtime/src/atn/ArrayPredictionContext.cpp#L55). That in turn would [test null pointers](https://github.com/antlr/antlr4/blob/master/runtime/Cpp/runtime/src/support/Arrays.h#L35). The nullptr in the parents array has existed for 6 years at least.